### PR TITLE
Support gradient checkpointing for finetuning billion-scale models on a single GPU + Loose the version bound of torch to be <=1.12

### PR DESCRIPTION
**Pull request by sxjscience**
*Description of changes:*
After the change, we are able to finetune the [mT5-XL](https://huggingface.co/google/mt5-xl) model that has **1.7B** parameters on a single NVIDIA G4 GPU (In AutoMM, we only use the T5-encoder like in [Sentence-T5](https://aclanthology.org/2022.findings-acl.146.pdf) so it's 1.7B.). Thus, users can now use billion-scale models like mT5-XL on **a single g4.2x instance** with AutoGluon Mulimodal.

To run the feature, use the example script: https://gist.github.com/sxjscience/b011460574c7e41c47a42ba86d026cbc

```
Auto select gpus: [0]
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]

  | Name              | Type                         | Params
-------------------------------------------------------------------
0 | model             | HFAutoModelForTextPrediction | 1.7 B 
1 | validation_metric | AUROC                        | 0     
2 | loss_func         | CrossEntropyLoss             | 0     
-------------------------------------------------------------------
3.2 M     Trainable params
1.7 B     Non-trainable params
1.7 B     Total params
6,692.434 Total estimated model params size (MB)

```

In addition, for training the model on multiple GPUs, we can just use the `fsdp` strategy supported in pytorch lightning:
https://gist.github.com/sxjscience/a86fc9c8e49b10af9495bc30322c3c26

- [x] Support gradient checkpointing: https://pytorch.org/docs/stable/checkpoint.html. If it is turned on, T5 in Huggingface seems to have some issues related to https://discuss.pytorch.org/t/checkpoint-with-no-grad-requiring-inputs-problem/19117/9. We fixed the issue by adding a DummyLayer.
- [x] Fix LoRA. The previous version has problems when used together gradient checkpointing. Also, we revised patterns to support regex.
- [x] Add new "pooling_mode" flag in `model.hf_text`. It now supports "cls" or "mean"
- [x] Turn on `track_grad_norm` in Trainer
- [x] New problem_type "classification"
- [x] Performance verification (since we are just using encoder, it may have worse performance in some benchmarking datasets, but the result is reasonable).

@Raldir @FANGAreNotGnu @Linuxdex @bryanyzhu @zhiqiangdon @gradientsky

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/install-cpu-source.rst [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
     # CPU version of pytorch has smaller footprint - see installation instructions in
     # pytorch documentation - https://pytorch.org/get-started/locally/
-    pip3 install "torch>=1.0,<1.12+cpu" -f https://download.pytorch.org/whl/cpu/torch_stable.html
+    pip3 install "torch>=1.0,<1.13+cpu" -f https://download.pytorch.org/whl/cpu/torch_stable.html
```
Comment (sxjscience) [inline docs/install-cpu-source.rst:8]: @Innixma @yinweisu I revised our installation guide so that users can install torch 1.12

Comment (Innixma) [reply docs/install-cpu-source.rst:8]: Are we still claiming support for torch 1.0? Should we tighten bounds?

Comment (sxjscience) [reply docs/install-cpu-source.rst:8]: How about just recommending the latest torch in our tutorial? I think we should work for torch>=1.9.

**Comment by wenming** [comment]
really nice, let's prioritize reviewing this.

File: docs/install-cpu-source.rst (hunk 1 above)
Comment (sxjscience) [reply docs/install-cpu-source.rst:8]: I checked the requirement from torchmetrics, it is currently asking for torch 1.11 in the stable version. Thus, we may also recommend using torch 1.11 in our website for now and upgrade to torch 1.12 once that is stable. Meanwhile, I think we can turn on torch 1.12 in our unittests.

File: multimodal/src/autogluon/multimodal/data/process_text.py [hunk 2]
```diff
@@ -243,13 +244,23 @@ def build_one_token_sequence(
             max_length = self.max_len - (len(text_tokens) + 1)
         else:
             max_length = self.max_len - 2
+        if self.eos_only:
+            # For EOS-only, the tokens will be combined as
+            # [Field1 Tokens] [EOS] [Field2 Tokens] [EOS] [Field3 Tokens] [EOS]
+            # Otherwise, the tokens will be combined as
+            # [CLS] [Field1 Tokens] [SEP] [Field2 Tokens] [SEP] [Field3 Tokens] [EOS]
+            max_length += 1
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/data/process_text.py:252]: Why do we need to increase the max length in the `eos_only` mode?

Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/data/process_text.py:252]: The number of special tokens will be reduced by 1, so the max length we can allocate for the text field tokens will be increased by 1.

File: multimodal/tests/unittests/test_backward_compatibility.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 
 from autogluon.multimodal import MultiModalPredictor, AutoMMPredictor
 from autogluon.multimodal.utils import download
-from datasets import AmazonReviewSentimentCrossLingualDataset
+from unittest_datasets import AmazonReviewSentimentCrossLingualDataset
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/test_backward_compatibility.py:9]: Does the name `datasets` cause confusion? Since it is can only be used in in this unit tests folder, I feel the previous name should be fine.

File: multimodal/src/autogluon/multimodal/configs/optimization/adamw.yaml [hunk 4]
```diff
@@ -18,13 +18,17 @@ optimization:
...
   lora:
     r: 8
     alpha: 8
     filter: # Default fine-tune only query and value attention weights, recommended in https://arxiv.org/abs/2106.09685
       - "query"
       - "value"
+      - "^q$"  # The filter can also be a regex. We will use re.match(filter, name) to determine if LoRA should be applied.
+      - "^v$"
+      - "^k$"
+      - "^o$"
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/configs/optimization/adamw.yaml:31]: Any examples how to use them?

File: multimodal/src/autogluon/multimodal/models/huggingface_text.py [hunk 5]
```diff
@@ -28,6 +28,15 @@
 logger = logging.getLogger(AUTOMM)
 
 
+class DummyLayer(nn.Module):
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/huggingface_text.py:31]: Better to move `DummyLayer` into utils.py. Need to keep this file clean.

File: multimodal/src/autogluon/multimodal/models/huggingface_text.py [hunk 6]
```diff
@@ -28,6 +28,15 @@
...
+    def __init__(self):
+        super().__init__()
+        self.dummy_bias = nn.Parameter(torch.ones(1, dtype=torch.float32))
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/huggingface_text.py:34]: Do we need to specify `dtype=torch.float32`? I think lightning can automatically handle this.

File: multimodal/tests/unittests/test_backward_compatibility.py (hunk 3 above)
Comment (sxjscience) [reply multimodal/tests/unittests/test_backward_compatibility.py:9]: Yes, it conflicts with the datasets packagez

File: multimodal/tests/unittests/test_predictor_advanced.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                      "optimization.efficient_finetune": efficient_finetuning,
+                      "optimization.lr_decay": 1.0,
+                      "optimization.learning_rate": 1e-03,
+                      "optimization.max_epochs": 1,
+                      "env.per_gpu_batch_size": 2,
+                      "env.num_workers": 0,
+                      "env.num_workers_evaluation": 0,
+                      "env.num_gpus": 1,
+                  })
+    predictions = predictor.predict(test_data)
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/test_predictor_advanced.py:59]: Do we need to test `verify_predictor_save_load()`?

File: multimodal/src/autogluon/multimodal/models/huggingface_text.py [hunk 8]
```diff
@@ -61,18 +72,37 @@ def __init__(
...
+            # Remove the decoder in T5
+            del self.model.decoder
+        else:
+            self.is_t5 = False
+
+        self.gradient_checkpointing = gradient_checkpointing
+        if gradient_checkpointing:
+            self.model.gradient_checkpointing_enable()
+            if self.is_t5:
+                self.dummy_layer = DummyLayer()
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/huggingface_text.py:97]: Is this gradient checkpointing issue unique to the `t5` model? If it also exist in other models, the dummy layer shouldn't condition only on `t5`.

File: multimodal/src/autogluon/multimodal/models/huggingface_text.py (hunk 5 above)
Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/models/huggingface_text.py:31]: We can consider to refactor it later. This is  a issue I met when finetuning T5 models but haven’t seen such issue in other backbones. We may remove the hack later.

File: multimodal/src/autogluon/multimodal/models/huggingface_text.py [hunk 9]
```diff
@@ -61,18 +72,37 @@ def __init__(
...
         self.checkpoint_name = checkpoint_name
         self.num_classes = num_classes
+
         self.model = AutoModel.from_pretrained(checkpoint_name)
+        if isinstance(self.model, T5PreTrainedModel):
+            self.is_t5 = True
+            # Remove the decoder in T5
+            del self.model.decoder
+        else:
+            self.is_t5 = False
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/huggingface_text.py:91]: Could we wrap this into a util function to detect special models? I guess `t5` is just one of them we need to deal with specially. The return of this util function is the model type.

File: multimodal/src/autogluon/multimodal/data/process_text.py (hunk 2 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/data/process_text.py:252]: Got it.

File: multimodal/src/autogluon/multimodal/models/huggingface_text.py (hunk 6 above)
Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/models/huggingface_text.py:34]: Lightning will set the dtype/device of the input to forward. For parameters, their dtype will be converted if we call functions like mod.half().

File: multimodal/src/autogluon/multimodal/models/huggingface_text.py (hunk 9 above)
Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/models/huggingface_text.py:91]: Yes, but we may do it later. Encoder-decoder models like  BART should also be specially treated.

Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/models/huggingface_text.py:91]: In fact, we may need to figure out the way for supporting applying the decoder in model training.

File: multimodal/src/autogluon/multimodal/configs/optimization/adamw.yaml (hunk 4 above)
Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/configs/optimization/adamw.yaml:31]: Basically some layers may be “xxx.q.yyy”. So it’s matching the “q” element here.

**Comment by sxjscience** [comment]
I'm still running some benchmarks and will share results afterwards.

**Review by bryanyzhu — COMMENTED** [review]
Great feature. Left a few minor comments.

File: multimodal/src/autogluon/multimodal/configs/environment/default.yaml [hunk 10]
```diff
@@ -11,4 +11,5 @@ env:
   fast_dev_run: False
   deterministic: False
   auto_select_gpus: True
+  track_grad_norm: 2  # Whether to check gradient norm. We can set it to 2 to check for gradient norm.
```
[the remark this answers is not in the document]
Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/configs/environment/default.yaml:14]: Yes, this is a newly added feature. It will track the gradient norms in Tensorboard. It helped me debug the LoRA + gradient checkpointing scenario.

[the remark this answers is not in the document]
Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/configs/environment/default.yaml:14]: track_grad_norm: https://pytorch-lightning.readthedocs.io/en/stable/common/debugging.html#inspect-gradient-norms

File: multimodal/src/autogluon/multimodal/models/huggingface_text.py (hunk 9 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/models/huggingface_text.py:91]: Maybe we need a config `uses_decoder`.

Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/models/huggingface_text.py:91]: I find that people will tend to follow previous design to add new stuff. By following this design, people will add more if-else patterns here.

File: multimodal/src/autogluon/multimodal/models/huggingface_text.py (hunk 5 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/models/huggingface_text.py:31]: We don't when Pytorch will fix this. Moving it to utils is still better. This may not be a good example for future developers to follow.

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 11]
```diff
@@ -1746,10 +1751,15 @@ def save(self, path: str, standalone: Optional[bool] = False):
             model_path = os.path.join(self._save_path, "model.ckpt")
             if os.path.isfile(model_path):
                 shutil.copy(model_path, path)
+            else:
+                # FIXME(?) Fix the saving logic
+                RuntimeError(f"Cannot find the model checkpoint in '{model_path}'. Have you removed the folder that "
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/predictor.py:1756]: @zhiqiangdon

File: multimodal/src/autogluon/multimodal/data/infer_types.py [hunk 12]
```diff
@@ -51,6 +62,7 @@ def is_categorical_column(
     else:
         if threshold is None:
             if is_label:
+                # TODO(?) The following logic will be problematic if the task is few-shot learning.
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/data/infer_types.py:65]: @Linuxdex @Raldir Will be related to few-shot learning. We should fix the logic in a future PR.

File: multimodal/src/autogluon/multimodal/models/huggingface_text.py (hunk 5 above)
Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/models/huggingface_text.py:31]: If I do not add the DummyLayer, finetuning mT5-XL on MRPC with the implementation can only get 0.70 ACC. As a reference, adding the dummy layer can get 0.872 ACC.

```
# Without Dummy
{'acc': 0.7009803921568627, 'f1': 0.8195266272189349, 'time_cost': 4389.08921289444}` 

# With Dummy
{"acc": 0.8725490196078431, "f1": 0.9047619047619048, "time_cost": 8712.896693944931}
```

Also, we need to wrap dummy layer after the embedding layer so it is not generally applicable.

File: multimodal/tests/unittests/test_predictor_advanced.py (hunk 7 above)
Comment (sxjscience) [reply multimodal/tests/unittests/test_predictor_advanced.py]: Tested.

**Comment by sxjscience** [comment]
I think the performance is reasonable.

**Review by bryanyzhu — APPROVED** [review]

**Review by zhiqiangdon — APPROVED** [review]
